docs/_includes/Statements.md

```diff
@@ -239,6 +239,153 @@ Note that the form
+   * The caller and callee are here shown with 2 output arguments each. In fact they each can have any (positive) number of output arguments, as long as the first one has a failure-compatible type.
+
+It is important to note the connection between the failure-compatible types used in the caller and callee. They do not have to be the same type, but they must be closely related, as it must be possible for the callee's `PropagateFailure` to return a value of the caller's failure-compatible type. In practice this means that one such failure-compatible type should be used for an entire program. If a Dafny program uses a library shared by multiple programs, the library should supply such a type and it should be used by all the client programs (and, effectively, all Dafny libraries). 
+It is also the case that it is inconvenient to mix types such as `Outcome` above and `VoidOutcome` below.
+
+### RHS with expression list
+
+Instead of a failure-returning method call on the RHS of the statement, the RHS can instead be a list of expressions. As for a `:=` statement, in this form, there must be the same number of expressions on the left and right sides of `:-`. The semantics is very similar to that in the previous subsection. 
+   * The first RHS expression must have a failure-compatible type.
+   * All the assignments of RHS expressions to LHS values except for the first value are made.
+   * If the first RHS value (say `r`) responds `true` to `r.IsFailure()`, then `r.PropagateFailure()` is assigned to the first output value of the _caller_ and the execution of the caller's body is ended.
+   * If the first RHS value (say `r`) responds `false` to `r.IsFailure()`, then `r.Extract()` is assigned to the first LHS value of the `:-` statement. Execution of the caller's body continues with the statement following the `:-` statement.
+
+### Types without Extract
+
+An alternative form of the above behavior allows the failure-compatible type to omit the `Extract` method, A typical example would be
+```
+datatype VoidOutcome =
+| Success
+| Failure(error: string)
+{
+  predicate method IsFailure() {
+    this.Failure?
+  }
+  function method PropagateFailure(): VoidOutcome
+    requires IsFailure()
+  {
+  Failure(this.error); // this is Outcome<U>.Failure(...)
+  }
+}
+```
+Such a type also counts as a failure-compatible type for the purposes of a `:-` statement. However, now there is no `Extract` method and no value carried along with `Success`.
+The same exceptional return semantics apply. The only difference is that the list of LHS values now must omit what would be the first l-value, as there is no extracted value for it to get. Thus there may be no LHS values at all.
+
+In current Dafny, this form of exceptional return requires that the caller and callee each have only one output value, namely a value of this special failure-compatible extract-omitting type.
+
+TODO: Say more when the above restriction is lifted.
```

Comment:
Is this TODO still applicable? Do we need another issue for it?

Reply:
Implementation completed.